src/k2/persistence/plog_service/PlogServer.cpp

```diff
@@ -0,0 +1,161 @@
+        return _handleRead(std::move(request));
+    });
+
+    RPC().registerRPCObserver<dto::PlogSealRequest, dto::PlogSealResponse>(dto::Verbs::K23SI_PERSISTENT_SEAL, [this](dto::PlogSealRequest&& request) {
+        return _handleSeal(std::move(request));
+    });
+    _plogMap.clear();
+
+    return seastar::make_ready_future<>();
+}
+
+seastar::future<std::tuple<Status, dto::PlogCreateResponse>>
+PlogServer::_handleCreate(dto::PlogCreateRequest&& request){
+    K2DEBUG("Received create request for " << request.plogId);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter != _plogMap.end()) {
+        return RPCResponse(Statuses::S530_PlogId_is_duplicated("plog id already exists"), dto::PlogCreateResponse());
+    }
+    _plogMap.insert(std::pair<String,PlogPage >(std::move(request.plogId), PlogPage()));
+    return RPCResponse(Statuses::S201_Created("plog created"), dto::PlogCreateResponse());
+};
+
+seastar::future<std::tuple<Status, dto::PlogAppendResponse>>
+PlogServer::_handleAppend(dto::PlogAppendRequest&& request){
+    K2DEBUG("Received append request for " << request.plogId << " with size" << request.payload.getSize() << " and offset " << request.offset);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), dto::PlogAppendResponse());
+    }
+    if (iter->second.sealed){
+         return RPCResponse(Statuses::S409_Conflict("plog is sealed"), dto::PlogAppendResponse());
+    }
+    if (iter->second.offset != request.offset){
+        return RPCResponse(Statuses::S400_Bad_Request("offset inconsistent"), dto::PlogAppendResponse());
+    }
+    if (iter->second.offset + request.payload.getSize() > PLOG_MAX_SIZE){
+         return RPCResponse(Statuses::S400_Bad_Request("exceeds pLog limit"), dto::PlogAppendResponse());
+    }
+
+    dto::PlogAppendResponse response;
+    response.offset = iter->second.offset + request.payload.getSize();
+    response.bytes_appended = request.payload.getSize();
+
+    iter->second.offset += request.payload.getSize();
+    iter->second.payload.copyFromPayload(request.payload, request.payload.getSize());
+    
+    return RPCResponse(Statuses::S200_OK("append scuccess"), std::move(response));
+};
+
+
+seastar::future<std::tuple<Status, dto::PlogReadResponse>>
+PlogServer::_handleRead(dto::PlogReadRequest&& request){
+    K2DEBUG("Received read request for " << request.plogId << " with offset " << request.offset);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), dto::PlogReadResponse());
+    }
+    if (iter->second.offset < request.offset + request.size){
+         return RPCResponse(Statuses::S400_Bad_Request("exceed the maximun length"), dto::PlogReadResponse());
+    }
+    
+    iter->second.payload.seek(request.offset);

```

Comment:
I think with my proposed change to implement a `shareRegion` api in the payload, you don't have to seek here. For completeness though, i wanted to point out that you have to remember the current payload offset, and then seek to it after you are done sharing so that the payload is ready to append new data.

Reply:
put the seek inside the shareRegion